src/gas_metering/mod.rs

```diff
@@ -711,22 +723,47 @@ fn generate_gas_counter(gas_global: u32) -> (Type, Function) {
 fn inject_counter<R: Rules>(
     instructions: &wasmparser::FunctionBody,
     rules: &R,
+    param_count: u32,
     gas_func: u32,
 ) -> Result<wasm_encoder::Function> {
-    let blocks = determine_metered_blocks(instructions, rules)?;
-    insert_metering_calls(instructions, blocks, gas_func)
+    let (blocks, metered_instrs) = determine_metered_blocks(instructions, rules)?;
+    insert_metering_calls(instructions, blocks, metered_instrs, param_count, gas_func)
 }
 
 // Then insert metering calls into a sequence of instructions given the block locations and costs.
 fn insert_metering_calls(
     func_body: &wasmparser::FunctionBody,
     blocks: Vec<MeteredBlock>,
+    instructions: Vec<MeteredInstruction>,
+    param_count: u32,
     gas_func: u32,
 ) -> Result<wasm_encoder::Function> {
-    let mut new_func = wasm_encoder::Function::new(copy_locals(func_body)?);
+    // collect value types on which we will be doing dynamic gas math.
+    // We need those for temp locals because wasm has no other way to duplicate stack items..
+
+    // todo in the future when we want to do linear gas cost on instructions where
+    //  the last parameter doesn't happen to always be i32 this will need to be smarter
+    let mut has_i32_temp = false;
+    for i in &instructions {
+        match i.param_type {
+            ValType::I32 => has_i32_temp = true,
+            _ => return Err(anyhow!("dynamic gas charges only supported on I32 params")),
+        }
+    }
```

Comment:
This code is weird.

- We're setting `has_i32_temp` each time.
- We assert that it's always true.
- We only push a local if... we have at least one instruction?

What am I missing? It looks like we should be:

1. Checking that `instructions.iter().all(|i| i.param_type == ValType::I32)`.
2. Returning early if we have no instructions.

Reply:
The main reason to not return early is that, in theory we may have a non-i32 instruction.. tho we can also check that when we build `Vec<MeteredInstruction>`, which saves iterations here

Reply:
This is written the way it is because initially I wanted to support multiple value types, but then looked at instructions we would want dynamic pricing for, and it turned out that we really only need i32